### PR TITLE
global: Use std::optional instead of boost::optional

### PR DESCRIPTION
**Pull request by FreddyFunk**
This PR depends on PR #1570 and continues @lioncash work. `boost::optional` dependency of yuzu codebase should be completely eliminated by now.

### CODE REVIEW

**Comment by bunnei** [comment]
I've merged #1570, so feel free to rebase. Thanks!

**Comment by lioncash** [comment]
The removal of references should be replaced with optionals containing `std::reference_wrapper<T>` (where T is the type to contain) rather than taking them by value, to preserve behavior

**Comment by FreddyFunk** [comment]
@bunnei  Since I have already merged #1570 in this branch a rebase is not necessary anymore.

@lioncash Fixed. Thanks for the hint!

File: src/core/crypto/key_manager.cpp [hunk 1]
```diff
@@ -890,7 +890,7 @@ void KeyManager::DeriveETicket(PartitionDataManager& data) {
 
     for (const auto& raw : res) {
         const auto pair = ParseTicket(raw, rsa_key);
-        if (pair == boost::none)
+        if (pair)
```
Comment (MerryMage) [inline src/core/crypto/key_manager.cpp]: `if (!pair)`

ditto throughout

File: src/core/crypto/key_manager.cpp [hunk 2]
```diff
@@ -661,8 +661,8 @@ void KeyManager::DeriveSDSeedLazy() {
         return;
 
     const auto res = DeriveSDSeed();
-    if (res != boost::none)
-        SetKey(S128KeyType::SDSeed, res.get());
+    if (res != std::nullopt)
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/crypto/key_manager.cpp]: Done

File: src/core/crypto/key_manager.cpp (hunk 1 above)
Comment (FreddyFunk) [reply src/core/crypto/key_manager.cpp]: Whoops, had a typo in the regular expression. Fixed now.

File: src/core/file_sys/content_archive.cpp [hunk 3]
```diff
@@ -306,18 +305,18 @@ bool NCA::ReadRomFSSection(const NCASectionHeader& section, const NCASectionTabl
...
         subsection_buckets.back().entries.push_back({size, {0}, 0});
 
-        boost::optional<Core::Crypto::Key128> key = boost::none;
+        std::optional<Core::Crypto::Key128> key = {};
         if (encrypted) {
             if (has_rights_id) {
                 status = Loader::ResultStatus::Success;
                 key = GetTitlekey();
-                if (key == boost::none) {
+                if (key) {
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/content_archive.cpp]: Done

File: src/core/file_sys/ips_layer.cpp [hunk 4]
```diff
@@ -103,12 +103,12 @@ VirtualFile PatchIPS(const VirtualFile& in, const VirtualFile& ips) {
             offset += sizeof(u16);
 
             const auto data = ips->ReadByte(offset++);
-            if (data == boost::none)
+            if (data)
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/ips_layer.cpp]: Done

File: src/core/file_sys/ips_layer.cpp [hunk 5]
```diff
@@ -103,12 +103,12 @@ VirtualFile PatchIPS(const VirtualFile& in, const VirtualFile& ips) {
...
                 return nullptr;
 
             if (real_offset + rle_size > in_data.size())
                 rle_size = static_cast<u16>(in_data.size() - real_offset);
-            std::memset(in_data.data() + real_offset, data.get(), rle_size);
+            std::memset(in_data.data() + real_offset, data.value(), rle_size);
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/ips_layer.cpp]: Done

File: src/core/file_sys/patch_manager.cpp [hunk 6]
```diff
@@ -281,12 +281,12 @@ std::map<std::string, std::string, std::less<>> PatchManager::GetPatchVersionNam
     } else {
         if (installed->HasEntry(update_tid, ContentRecordType::Program)) {
             const auto meta_ver = installed->GetEntryVersion(update_tid);
-            if (meta_ver == boost::none || meta_ver.get() == 0) {
+            if (meta_ver || meta_ver.value() == 0) {
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/patch_manager.cpp]: Can be reduced to `(meta_ver.value_or(0) == 0)`

File: src/core/file_sys/patch_manager.cpp [hunk 7]
```diff
@@ -281,12 +281,12 @@ std::map<std::string, std::string, std::less<>> PatchManager::GetPatchVersionNam
...
                 out.insert_or_assign("Update", "");
             } else {
                 out.insert_or_assign(
                     "Update",
-                    FormatTitleVersion(meta_ver.get(), TitleVersionFormat::ThreeElements));
+                    FormatTitleVersion(meta_ver.value(), TitleVersionFormat::ThreeElements));
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/patch_manager.cpp]: Done

File: src/core/file_sys/registered_cache.cpp [hunk 8]
```diff
@@ -283,17 +283,17 @@ bool RegisteredCache::HasEntry(RegisteredCacheEntry entry) const {
 
 VirtualFile RegisteredCache::GetEntryUnparsed(u64 title_id, ContentRecordType type) const {
     const auto id = GetNcaIDFromMetadata(title_id, type);
-    if (id == boost::none)
+    if (id)
         return nullptr;
 
-    return GetFileAtID(id.get());
+    return GetFileAtID(id.value());
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/registered_cache.cpp]: That's great, done

File: src/core/file_sys/registered_cache.cpp [hunk 9]
```diff
@@ -302,15 +302,15 @@ boost::optional<u32> RegisteredCache::GetEntryVersion(u64 title_id) const {
...
         return yuzu_meta_iter->second.GetTitleVersion();
 
-    return boost::none;
+    return {};
 }
 
 VirtualFile RegisteredCache::GetEntryRaw(u64 title_id, ContentRecordType type) const {
     const auto id = GetNcaIDFromMetadata(title_id, type);
-    if (id == boost::none)
+    if (id)
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/registered_cache.cpp]: Done

File: src/core/file_sys/registered_cache.cpp [hunk 10]
```diff
@@ -364,20 +364,20 @@ std::vector<RegisteredCacheEntry> RegisteredCache::ListEntries() const {
...
+    std::optional<u64> title_id) const {
     std::vector<RegisteredCacheEntry> out;
     IterateAllMetadata<RegisteredCacheEntry>(
         out,
         [](const CNMT& c, const ContentRecord& r) {
             return RegisteredCacheEntry{c.GetTitleID(), r.type};
         },
         [&title_type, &record_type, &title_id](const CNMT& c, const ContentRecord& r) {
-            if (title_type != boost::none && title_type.get() != c.GetType())
+            if (title_type != std::nullopt && title_type.value() != c.GetType())
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/registered_cache.cpp]: I think you mean `if (title_type && *title_type != c.GetType())`

File: src/core/file_sys/registered_cache.cpp [hunk 11]
```diff
@@ -468,12 +468,12 @@ InstallResult RegisteredCache::RawInstallNCA(std::shared_ptr<NCA> nca, const Vfs
     // game is massive), we're going to cheat and only hash the first MB of the NCA.
     // Also, for XCIs the NcaID matters, so if the override id isn't none, use that.
     NcaID id{};
-    if (override_id == boost::none) {
+    if (override_id) {
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/registered_cache.cpp]: That's great, done

File: src/core/file_sys/registered_cache.cpp [hunk 12]
```diff
@@ -543,14 +543,14 @@ bool RegisteredCacheUnion::HasEntry(RegisteredCacheEntry entry) const {
     return HasEntry(entry.title_id, entry.type);
 }
 
-boost::optional<u32> RegisteredCacheUnion::GetEntryVersion(u64 title_id) const {
+std::optional<u32> RegisteredCacheUnion::GetEntryVersion(u64 title_id) const {
     for (const auto& c : caches) {
         const auto res = c->GetEntryVersion(title_id);
-        if (res != boost::none)
+        if (res != std::nullopt)
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/registered_cache.cpp]: Done

File: src/core/file_sys/registered_cache.cpp [hunk 13]
```diff
@@ -619,11 +619,11 @@ std::vector<RegisteredCacheEntry> RegisteredCacheUnion::ListEntriesFilter(
                 return RegisteredCacheEntry{c.GetTitleID(), r.type};
             },
             [&title_type, &record_type, &title_id](const CNMT& c, const ContentRecord& r) {
-                if (title_type != boost::none && title_type.get() != c.GetType())
+                if (title_type != std::nullopt && title_type.value() != c.GetType())
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/file_sys/registered_cache.cpp]: Done

File: src/core/hle/service/nvflinger/nvflinger.cpp [hunk 14]
```diff
@@ -143,7 +143,7 @@ void NVFlinger::Compose() {
             continue;
         }
 
-        auto& igbp_buffer = buffer->igbp_buffer;
+        auto& igbp_buffer = buffer.value().get().igbp_buffer;
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/hle/service/nvflinger/nvflinger.cpp]: Done

File: src/core/hle/service/vi/vi.cpp [hunk 15]
```diff
@@ -506,9 +507,9 @@ class IHOSBinderDriver final : public ServiceFramework<IHOSBinderDriver> {
             IGBPDequeueBufferRequestParcel request{ctx.ReadBuffer()};
             const u32 width{request.data.width};
             const u32 height{request.data.height};
-            boost::optional<u32> slot = buffer_queue->DequeueBuffer(width, height);
+            std::optional<u32> slot = buffer_queue->DequeueBuffer(width, height);
 
-            if (slot != boost::none) {
+            if (slot != std::nullopt) {
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/core/hle/service/vi/vi.cpp]: Done

File: src/video_core/macro_interpreter.cpp [hunk 16]
```diff
@@ -44,10 +44,10 @@ bool MacroInterpreter::Step(const std::vector<u32>& code, bool is_delay_slot) {
     pc += 4;
 
     // Update the program counter if we were delayed
-    if (delayed_pc != boost::none) {
+    if (delayed_pc != std::nullopt) {
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/video_core/macro_interpreter.cpp]: Done

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 17]
```diff
@@ -143,7 +143,7 @@ class ControlFlowAnalyzer {
         for (u32 offset = begin; offset != end && offset != PROGRAM_END; ++offset) {
             const Instruction instr = {program_code[offset]};
             if (const auto opcode = OpCode::Decode(instr)) {
-                switch (opcode->GetId()) {
+                switch (opcode.value().get().GetId()) {
```
[the remark this answers is not in the document]
Comment (FreddyFunk) [reply src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Done

**Comment by FreddyFunk** [comment]
Rebased

**Comment by bunnei** [comment]
Sorry - needs one more rebase then I'll merge! Thanks.

**Comment by FreddyFunk** [comment]
Rebased

**Pull request merged**